[16.0] [ADD] Add bases of volunteer module

aydrpm opened the pull request:
## Description

Foundation of the volunteer module.

This PR includes:
- Module creation
- Security groups
- Menu
- Models and views for:
  - Shift
  - Volunteer
  - Category
  - Tag (with colors)
  - Type

Notes:
-  The "CheckViolation" log "volunteer_shift_max_volunteer_nb_is_positive" appears during the "test_reduce_max_volunteer_equals_zero"  test.
This log should be muted, but the test still passes.

### Planned for Deliverable 2

Filters:
- Smart buttons on the volunteer form
  - All shifts
  - Future shifts (from now)
- Kanban view filtered on future shifts
- Other filters to define

Access Rights:
- Check creation rights in notebooks (tag, type, etc.)
- Manage shift state transitions (status bar)
- Check delete permissions in the action menu

Field and UI:
- Review field options
- Add chatter tracking
- Shift state color display

Constraints:
- General checks, including uniqueness constraints
- Prevent registrations on draft or canceled shifts
- Prevent canceled → confirmed transitions for shifts and participations  
  (except for admin, with chatter logs)

Notifications:
- Send mail after participation state change

### Future Deliverables

- Tab on the volunteer form for recurring shifts
- Automatically update participations when a shift is canceled
- Compute the "is_regular" field


## Odoo task (if applicable)



## Checklist before approval

- [ ] Tests are present (or not needed).
- [ ] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [ ] (If a new module) Moving this to OCA has been considered.

aydrpm committed:
[ADD] volunteer

aydrpm committed:
Minor fixes shift models:
- Added timezone selection field with user tz by default
- Fixed dependencies _compute_name
- Added end day of shift in the computed name
- Added "draft" state; correction "new" to "confirmed"

aydrpm committed:
Modification of the editable name field

Commenting on the work in progress on the calculated field

aydrpm committed:
Add shift states management in shift form header

- Add statusbar
- Add confirm/cancel shift button

aydrpm committed:
Minor fixes

- shift name required=True
- fix tests
- cleaning

aydrpm committed:
Minor changes

- rename field timezone tz
- add docstring test

remytms commented on volunteer/data/volunteer.shift.category.csv:
Can you switch from CSV data to XML data ? Also can you put these file into a `demo` folder instead of in the `data` folder. Thanks.

aydrpm replied:
Done.

remytms commented on volunteer/models/volunteer_shift.py:
```diff
@@ -0,0 +1,84 @@
+
+    @api.model
+    def _tz_get(self):
+        return [(tz, tz) for tz in pytz.all_timezones]
```
No need to rewrite this. Use `from odoo.addons.base.models.res_partner import _tz_get` as it is done in `event` module.

aydrpm replied:
Done.

aydrpm committed:
Improved Shift form layout

remytms commented on volunteer/models/volunteer_shift.py:
```diff
@@ -0,0 +1,84 @@
+    #         if (record.name == False or record.name is "") or
+    #                                   re.match(pattern,record.name) != None:
+    #             record.name = computed_name
+    #     return True
```
Use same logic for dates than in the `event` module: https://github.com/odoo/odoo/blob/16.0/addons/event/models/event_event.py#L195-L204

aydrpm replied:
Done

remytms commented on volunteer/security/volunteer_security.xml:
```diff
@@ -0,0 +1,56 @@
+            <field
+                name="category_id"
+                ref="base.module_category_cooperative_management"
+            />
```
Create your own category here. Like `base.module_volunteer_management`.

aydrpm replied:
Done.

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,120 @@
+            }
+        )
+
+    def test_shift_create(self):
+        self.assertEqual(self.shift1.name, "Test")
+        self.assertEqual(self.shift1.state, "canceled")
+        self.assertEqual(self.shift1.timezone, "Europe/Brussels")
+        self.assertEqual(self.shift1.max_volunteer_nb, 2)
+        self.assertEqual(self.shift1.company_id, self.env.user.company_id)
```
You can remove this test in a new commit.

aydrpm replied:
Done.

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,120 @@
+        self.assertEqual(self.shift1.max_volunteer_nb, 2)
+        self.assertEqual(self.shift1.company_id, self.env.user.company_id)
+
+    def test_shift_create_default_values(self):
+        self.assertEqual(self.shift_default.name, "Test")
+        self.assertEqual(self.shift_default.state, "draft")
+        self.assertEqual(self.shift_default.timezone, self.env.user.tz)
+        self.assertEqual(self.shift_default.max_volunteer_nb, 1)
+        self.assertEqual(self.shift_default.company_id, self.env.user.company_id)
```
You can remove this test.

aydrpm replied:
Done.

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,120 @@
+        self.assertEqual(self.shift_default.max_volunteer_nb, 1)
+        self.assertEqual(self.shift_default.company_id, self.env.user.company_id)
+
+    def test_shift_relations(self):
+        self.assertEqual(self.shift_default.type_id.name, "TypeTest")
+        self.assertEqual(self.shift_default.tag_ids.name, "TagTest")
+        self.assertEqual(self.shift_default.category_id.name, "CategoryTest")
```
You can remove this test.

aydrpm replied:
Done.

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,120 @@
+        self.assertEqual(self.shift_default.tag_ids.name, "TagTest")
+        self.assertEqual(self.shift_default.category_id.name, "CategoryTest")
+
+    def test_shift_time_values(self):
+        self.assertEqual(
+            self.shift_default.start_time, datetime.datetime(2025, 12, 24, 10, 5)
+        )
+        self.assertEqual(
+            self.shift_default.end_time, datetime.datetime(2025, 12, 24, 12, 5)
+        )
```
You can remove this test.

aydrpm replied:
Done.

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,120 @@
+            self.shift_default.end_time, datetime.datetime(2025, 12, 24, 12, 5)
+        )
+
+    def test_convert_naive_date(self):
+        aware_date = _convert_naive_to_str_aware_date(
+            self.shift_default.start_time, "Europe/Brussels", "%Y/%m/%d_%H:%M"
+        )
+        self.assertEqual(aware_date, "2025/12/24_11:05")
```
I think this will not be needed anymore if you change to the `event` way of managing dates. Also, it can be good, to add a test that `compute` method for dates and with timezone are ok.

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,120 @@
+        )
+        self.assertEqual(aware_date, "2025/12/24_11:05")
+
+        ###########################################
+        #       Test Computed field name          #
+        #       Work in Progress for addition     #
+        #       in a future another modules       #
+        ###########################################
+
+    # def test_shift_compute_name_custom_tz(self):
+    #     formated_start_time = _convert_naive_to_str_aware_date(
+    #         self.shift1.start_time, "Europe/Brussels", "%Y/%m/%d_%H:%M"
+    #     )
+    #     formated_end_time = _convert_naive_to_str_aware_date(
+    #         self.shift1.end_time, "Europe/Brussels", "%H:%M"
+    #     )
+    #     self.assertEqual(
+    #         self.shift1.name,
+    #         f"{formated_start_time}-{formated_end_time}_CategoryTest_TypeTest",
+    #     )
+    #
+    # def test_shift_compute_name_default_tz(self):
+    #     start_time = datetime.datetime(2025, 12, 24, 10, 5).astimezone(
+    #         pytz.timezone(self.shift_default.timezone)
+    #     )
+    #     formated_start_time = datetime.datetime.strftime(start_time, "%Y/%m/%d_%H:%M")
+    #     end_time = datetime.datetime(2025, 12, 24, 12, 5).astimezone(
+    #         pytz.timezone(self.shift_default.timezone)
+    #     )
+    #     formated_end_time = datetime.datetime.strftime(end_time, "%H:%M")
+    #     self.assertEqual(
+    #         self.shift_default.name,
+    #         f"{formated_start_time}-{formated_end_time}_CategoryTest_TypeTest",
+    #     )
```
Just remove that in a new commit. If you need to get this code back, you can look in your git log.

aydrpm replied:
Done.

remytms commented on volunteer/tests/test_volunteer_shift_category.py:
```diff
@@ -0,0 +1,19 @@
+from odoo.tests import common
+
+
+class TestCategory(common.TransactionCase):
+
+    def setUp(self, *args, **kwargs):
+        super().setUp(*args, **kwargs)
+        self.Category = self.env["volunteer.shift.category"]
+        self.category1 = self.Category.create(
+            {
+                "name": "CategoryTest",
+                "description": "Category pour autotests",
+            }
+        )
+
+    def test_category_create(self):
+        self.assertEqual(self.category1.name, "CategoryTest")
+        self.assertEqual(self.category1.description, "Category pour autotests")
+        self.assertEqual(self.category1.company_id, self.env.user.company_id)
```
Remove this file, we will not test categories. :)

aydrpm replied:
Done.

remytms commented on volunteer/tests/test_volunteer_shift_tag.py:
```diff
@@ -0,0 +1,19 @@
+from odoo.tests import common
+
+
+class TestShift(common.TransactionCase):
+
+    def setUp(self, *args, **kwargs):
+        super().setUp(*args, **kwargs)
+        self.Tag = self.env["volunteer.shift.tag"]
+        self.tag1 = self.Tag.create(
+            {
+                "name": "TagTest",
+                "description": "Tag pour autotests",
+            }
+        )
+
+    def test_tag_create(self):
+        self.assertEqual(self.tag1.name, "TagTest")
+        self.assertEqual(self.tag1.description, "Tag pour autotests")
+        self.assertEqual(self.tag1.company_id, self.env.user.company_id)
```
Same remove this file. We will not test tags. :)

aydrpm replied:
Done.

remytms commented on volunteer/tests/test_volunteer_shift_type.py:
```diff
@@ -0,0 +1,19 @@
+from odoo.tests import common
+
+
+class TestShift(common.TransactionCase):
+
+    def setUp(self, *args, **kwargs):
+        super().setUp(*args, **kwargs)
+        self.Type = self.env["volunteer.shift.type"]
+        self.type1 = self.Type.create(
+            {
+                "name": "TypeTest",
+                "description": "Type pour autotests",
+            }
+        )
+
+    def test_type_create(self):
+        self.assertEqual(self.type1.name, "TypeTest")
+        self.assertEqual(self.type1.description, "Type pour autotests")
+        self.assertEqual(self.type1.company_id, self.env.user.company_id)
```
You can also remove this file.

aydrpm replied:
Done.

remytms commented on volunteer/views/shift_view.xml:
```diff
@@ -0,0 +1,62 @@
+            <form string="Shift">
+                <header>
+                    <field name="state" invisible="True" />
+                    <button
+                        name="button_confirm_shift"
+                        type="object"
+                        string="Confirm Shift"
+                        states="draft"
+                        class="oe_highlight"
+                    />
+                     <button
+                        name="button_cancel_shift"
+                        type="object"
+                        string="Cancel Shift"
+                        states="confirmed"
+                    />
```
I’m wondering if these buttons are needed. User can switch status directly from statusbar.

Buttons are needed when operation should be performed when the status of the object changes.

We can discuss that tomorrow. :)

aydrpm replied:
Done.

remytms commented on volunteer/views/shift_view.xml:
```diff
@@ -0,0 +1,62 @@
+                <sheet>
+                    <group name="group_top">
+                        <group name="group_left">
+                            <field name="name" />
```
Put name outside of the group left. So that the name is in large at the top.

aydrpm replied:
Done.

remytms commented on volunteer/views/shift_view.xml:
```diff
@@ -0,0 +1,62 @@
+                    <group name="group_top">
+                        <group name="group_left">
+                            <field name="name" />
+                            <field name="start_time" />
+                            <field name="end_time" />
```
Looks how dates are managed in `event` module: https://github.com/odoo/odoo/blob/16.0/addons/event/views/event_event_views.xml#L43-L48

aydrpm replied:
Done.

remytms commented on volunteer/views/shift_view.xml:
```diff
@@ -0,0 +1,62 @@
+        <field name="arch" type="xml">
+            <tree>
+                <field name="state" />
+                <field name="name" />
```
I think I will put the name before the status. Put the status at the end. It’s like this in other object list view. :)

aydrpm replied:
Done.

remytms commented on volunteer/views/shift_view.xml:
```diff
@@ -0,0 +1,62 @@
+                <field name="type_id" />
+                <field name="tag_ids" widget="many2many_tags" />
+                <field name="company_id" />
+                <field name="timezone" />
```
There is a way to put optional fields in list views. Timezone can be one of these fields.

To make them optional use the `optional` attribute with value `show` or `hide` to set the default display.

Looks example here: https://github.com/odoo/odoo/blob/16.0/addons/event/views/event_event_views.xml#L128

aydrpm replied:
Done.

remytms commented on volunteer/views/volunteer_menu.xml:
```diff
@@ -0,0 +1,61 @@
+        action="action_volunteer_shift"
+    />
+
+    <!-- Menu item to open Configuration menu -->
+    <menuitem
+        id="menu_volunteer_configuration"
+        name="Configuration"
+        parent="menu_volunteer"
+    />
```
I think that configuration menu should be visible only by the users in the admin group of this modules. Other should not see it.

aydrpm replied:
Done.

remytms reviewed (changes requested):
Great job ! :) Nice code. Pleasant to read.

I made some changes request. :)

aydrpm committed:
Removal of computed field tests commented

aydrpm committed:
Change of module category

aydrpm committed:
Removed rewriting of _tz_get

aydrpm committed:
Removed unnecessary tests

aydrpm committed:
Restriction access rights config menu to admins

aydrpm committed:
Minor shift view fixes

- Removed unnecessary actions buttons (form)
- Statusbar clickable (form)
- State field move at the right end (list)
- Timezone fields modify to optional hidden (list)

aydrpm committed:
Modification demo data csv to xml

aydrpm committed:
Correction logic management dates

aydrpm committed:
Work in Progress imp volunteer and participation

aydrpm committed:
Added xml headers

aydrpm committed:
Edit layout shift_view

+ rename views prefixed by shift
+ add required max_volunteer_nb

aydrpm committed:
[ADD] Volunteer and shift_participation models

- add views
- add ACLs
- add menus
- add mail dependencies for partner inherit

aydrpm committed:
Del computed fields is_regular

- Correction of participation-volunteer relationship

aydrpm committed:
Add registration type "manual" in participation

- set "manual" by default
- del dedicated participation view and menu
- add computed name participation

aydrpm committed:
Customisation shift views and add chatter

aydrpm committed:
Add shift kanban view and max volunteer constraint

- Stages Management for kanban view
- Add stages data
- Add check_remaining_slots constraint
- Add custom participation form (trigger on notebook shift)
- Change display name replaced volunteer "Replaced By"

aydrpm committed:
Work in progress volunteer views (form, kanban)

aydrpm committed:
Add coordinators management and cleanup

- added inline editing for coordinators and volunteers in shift notebook
- made registration_state required
- removed compute_name from participation
- removed replaced_volunteer field
- removed state from volunteer address
- removed custom view for participation form

aydrpm committed:
Add demo data (shift, participation, volunteer)

aydrpm committed:
Add participation cancellation_date computed

aydrpm committed:
Shift improvement constraints and layout view

- add fields is_one_day and remaining_slot computed
- add constraints max_volunteer
- fix sql constraints max_volunteer_nb_positive
- add function can_add_participations

- Improved layout shift view

- coordinator_ids changed to many2one coordinator_id
 - del coordinator_lines in notebook
 - add coordinator selection fields in form shift and moved the max_volunteer field

aydrpm committed:
Fix demo data

aydrpm committed:
Add constrains remaining_slot shift_participation

- fix typo 'cancelled' to 'canceled'

aydrpm committed:
Improve kanban views (shift and volunteer)

aydrpm committed:
Add notebook participation to volunteer view

- fix layout image form

aydrpm committed:
Fix and add test

aydrpm committed:
Improve shift kanban layout with CSS

aydrpm committed:
General cleaning: docstrings, naming, readability

aydrpm committed:
Add coordinator to shift demo

aydrpm committed:
Assign shift stages in shift demo data

- fix typo record id stage_canceled

aydrpm committed:
Fix volunteer_ids field in shift to computed field

remytms commented on volunteer/__manifest__.py:
```diff
@@ -0,0 +1,45 @@
+    "category": "Volunteer management",
+    "website": "https://github.com/beescoop/Obeesdoo",
+    "author": "Coop IT Easy SC",
+    "maintainers": ["aydrpm"],
```
Can you also add me `remytms` as a maintainer of this module. :)

aydrpm replied:
Done

remytms commented on volunteer/__manifest__.py:
```diff
@@ -0,0 +1,45 @@
+    "license": "AGPL-3",
+    "application": True,
+    "depends": ["mail"],
+    "excludes": [],
```
You can remove this key (`excludes`) from the manifest as it is empty.

aydrpm replied:
Done

remytms commented on volunteer/__manifest__.py:
```diff
@@ -0,0 +1,45 @@
+        "demo/shift_demo.xml",
+        "demo/shift_participation_demo.xml",
+    ],
+    "qweb": [],
```
Also remove this key (`qweb`) from manifest.

aydrpm replied:
Done

remytms commented on volunteer/data/shift_stage.xml:
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="utf-8" ?>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
```
You can add this to all the xml files.

aydrpm replied:
Done

remytms commented on volunteer/demo/shift_category_demo.xml:
```diff
@@ -0,0 +1,22 @@
+      <record model="volunteer.shift.category" id="shift_category_d">
+        <field name="name">Week D</field>
+        <field name="description">Shift Week D</field>
+    </record>
```
```suggestion
    <record model="volunteer.shift.category" id="shift_category_b">
        <field name="name">Week B</field>
        <field name="description">Shift Week B</field>
    </record>

    <record model="volunteer.shift.category" id="shift_category_c">
        <field name="name">Week C</field>
        <field name="description">Shift Week C</field>
    </record>

    <record model="volunteer.shift.category" id="shift_category_d">
        <field name="name">Week D</field>
        <field name="description">Shift Week D</field>
    </record>
```

aydrpm replied:
Done

remytms commented on volunteer/demo/shift_tag_demo.xml:
```diff
@@ -0,0 +1,22 @@
+      <record model="volunteer.shift.tag" id="shift_tag_d">
+        <field name="name">Tag D</field>
+        <field name="description">Shift Tag D</field>
+    </record>
```
```suggestion
    <record model="volunteer.shift.tag" id="shift_tag_b">
        <field name="name">Tag B</field>
        <field name="description">Shift Tag B</field>
    </record>

    <record model="volunteer.shift.tag" id="shift_tag_c">
        <field name="name">Tag C</field>
        <field name="description">Shift Tag C</field>
    </record>

    <record model="volunteer.shift.tag" id="shift_tag_d">
        <field name="name">Tag D</field>
        <field name="description">Shift Tag D</field>
    </record>
```

aydrpm replied:
Done

remytms commented on volunteer/demo/shift_type_demo.xml:
```diff
@@ -0,0 +1,22 @@
+      <record model="volunteer.shift.type" id="shift_type_delicatessen_cut">
+        <field name="name">Delicatessen Cutting</field>
+        <field name="description">Shift Delicatessen Cutting</field>
+    </record>
```
```suggestion
    <record model="volunteer.shift.type" id="shift_type_closing">
        <field name="name">Store Closing</field>
        <field name="description">Shift Closing Store</field>
    </record>

    <record model="volunteer.shift.type" id="shift_type_cheese_cut">
        <field name="name">Cheese Cutting</field>
        <field name="description">Shift Cheese Cutting</field>
    </record>

    <record model="volunteer.shift.type" id="shift_type_delicatessen_cut">
        <field name="name">Delicatessen Cutting</field>
        <field name="description">Shift Delicatessen Cutting</field>
    </record>
```

aydrpm replied:
Done

remytms commented on volunteer/models/__init__.py:
```diff
@@ -0,0 +1,8 @@
+from . import volunteer_shift
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from . import volunteer_shift
```

aydrpm replied:
Done

remytms commented on volunteer/models/res_partner.py:
```diff
@@ -0,0 +1,9 @@
+from odoo import fields, models
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from odoo import fields, models
```

aydrpm replied:
Done

aydrpm committed:
Add tag color and formless creation in list view

- formless creation/edition and company_id hide (categories, types, tags)

remytms commented on volunteer/models/volunteer_shift.py:
```diff
@@ -0,0 +1,202 @@
+from odoo import api, fields, models
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from odoo import api, fields, models
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift.py:
```diff
@@ -0,0 +1,202 @@
+    @api.model
+    def _default_stage_id(self):
+        Stage = self.env["volunteer.shift.stage"]
+        return Stage.search([("state", "=", "draft")], limit=1)
```
I would suggest using the element in the data instead of doing a search, because this can brings trouble if user creates new stages:
```suggestion
        return self.env.ref("volunteer.stage_draft")
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift.py:
```diff
@@ -0,0 +1,202 @@
+    # Classification fields
+
+    type_id = fields.Many2one("volunteer.shift.type", "Type", required=True)
+    category_id = fields.Many2one("volunteer.shift.category", "Category")
```
```suggestion
    type_id = fields.Many2one("volunteer.shift.type", string="Type", required=True)
    category_id = fields.Many2one("volunteer.shift.category", string="Category")
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift.py:
```diff
@@ -0,0 +1,202 @@
+    def _unique_volunteer_participation(self):
+        """Check that a volunteer can only be registered once per shift."""
+
+        for shift in self:
```
```suggestion
        """Check that a volunteer can only be registered once per shift."""
        for shift in self:
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift.py:
```diff
@@ -0,0 +1,202 @@
+    def _set_tz_context(self):
+        """Set the timezone context for the shift."""
+
+        self.ensure_one()
```
```suggestion
        """Set the timezone context for the shift."""
        self.ensure_one()
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift.py:
```diff
@@ -0,0 +1,202 @@
+        - nb_confirmed_participation: number of confirmed participation
+        """
+
+        self.ensure_one()
```
```suggestion
        """Get the shift booking status by returning a dictionary of :
        - can_accept_participation: boolean
        - confirmed_participation: recordset of confirmed participation
        - nb_confirmed_participation: number of confirmed participation
        """
        self.ensure_one()
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_volunteer.py:
```diff
@@ -0,0 +1,27 @@
+        "volunteer.shift.participation", "volunteer_id", "participation"
+    )
+
+    is_regular = fields.Boolean(
```
```suggestion
    )
    is_regular = fields.Boolean(
```

remytms commented on volunteer/security/volunteer_security.xml:
```diff
@@ -0,0 +1,56 @@
+        <record id="volunteer_group_user" model="res.groups">
+            <field name="name">Volunteer</field>
+            <field name="category_id" ref="base.module_category_volunteer_management" />
+            <field name="implied_ids" eval="[(4,ref('base.group_user'))]" />
```
```suggestion
            <field name="implied_ids" eval="[(4, ref('base.group_user'))]" />
```

aydrpm replied:
Done

remytms commented on volunteer/security/volunteer_security.xml:
```diff
@@ -0,0 +1,56 @@
+        <record id="volunteer_group_manager" model="res.groups">
+            <field name="name">Manager</field>
+            <field name="category_id" ref="base.module_category_volunteer_management" />
+            <field name="implied_ids" eval="[(4,ref('volunteer_group_user'))]" />
```
```suggestion
            <field name="implied_ids" eval="[(4, ref('volunteer_group_user'))]" />
```

aydrpm replied:
Done

remytms commented on volunteer/security/volunteer_security.xml:
```diff
@@ -0,0 +1,56 @@
+        <record id="volunteer_group_admin" model="res.groups">
+            <field name="name">Admin</field>
+            <field name="category_id" ref="base.module_category_volunteer_management" />
+            <field name="implied_ids" eval="[(4,ref('volunteer_group_manager'))]" />
```
```suggestion
            <field name="implied_ids" eval="[(4, ref('volunteer_group_manager'))]" />
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift_category.py:
```diff
@@ -0,0 +1,15 @@
+from odoo import fields, models
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from odoo import fields, models
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift_participation.py:
```diff
@@ -0,0 +1,69 @@
+from odoo import api, fields, models
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from odoo import api, fields, models
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift_stage.py:
```diff
@@ -0,0 +1,16 @@
+from odoo import fields, models
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from odoo import fields, models
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift_tag.py:
```diff
@@ -0,0 +1,23 @@
+from random import randint
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from random import randint
```

aydrpm replied:
Done

remytms commented on volunteer/models/volunteer_shift_type.py:
```diff
@@ -0,0 +1,15 @@
+from odoo import fields, models
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from odoo import fields, models
```

aydrpm replied:
Done

remytms commented on volunteer/demo/shift_category_demo.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/demo/shift_demo.xml:
```diff
@@ -0,0 +1,47 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/demo/shift_participation_demo.xml:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/demo/shift_tag_demo.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/demo/shift_type_demo.xml:
```diff
@@ -0,0 +1,22 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/demo/volunteer_demo.xml:
```diff
@@ -0,0 +1,42 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/security/volunteer_security.xml:
```diff
@@ -0,0 +1,56 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/security/volunteer_security.xml:
```diff
@@ -0,0 +1,56 @@
+            <field name="model_id" ref="model_volunteer_volunteer" />
+            <field name="groups" eval="[(4, ref('volunteer_group_user'))]" />
+        </record>
+    </data>
```
I know we have talk about that. But I’m wondering why just `ir.model.access.csv` is not enough to manage right on these models.

It is ok to let that like this, because you may come back later on that to add real rules for groups.

aydrpm replied:
You're right, it's not needed here.
I've removed the access rules for now and can refer back to this commit in the future if needed.

remytms commented on volunteer/models/volunteer_volunteer.py:
```diff
@@ -0,0 +1,27 @@
+from odoo import fields, models
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from odoo import fields, models
```

aydrpm replied:
Done

remytms commented on volunteer/tests/__init__.py:
```diff
@@ -0,0 +1,3 @@
+from . import test_volunteer_common
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from . import test_volunteer_common
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_common.py:
```diff
@@ -0,0 +1,64 @@
+from odoo.tests import common
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from odoo.tests import common
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+from psycopg2.errors import CheckViolation
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from psycopg2.errors import CheckViolation
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        shift created with tz UTC
+        expected is_one_day = True"""
+
+        shift_utc = self.Shift.create(
```
```suggestion
        """Test if a shift is one day or not"""
        shift_utc = self.Shift.create(
```

The comment is equivalent to the code so it’s not needed.

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        # shift created with tz UTC
+        # expected is_one_day = False
+
+        shift_utc = self.Shift.create(
```
```suggestion
        shift_utc = self.Shift.create(
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        """Test that the volunteer_ids are correctly computed
+        2 participation confirmed, 1 participation canceled"""
+
+        self.Participation.create(
```
```suggestion
        """Test that the volunteer_ids are correctly computed"""
        # There is one participation `confirmed` already defined in setUp()
        self.Participation.create(
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        UTC end_time = 2027-06-06 23:00:01
+        expected end_time_located = 2027-06-07 00:00:01"""
+
+        self.assertFalse(self.shift_utc_plus_1.is_one_day)
```
```suggestion
    def test_compute_time_located(self):
        """Test that the start_time_located and end_time_located
        are correctly computed with specific timezone"""
        # shift created with timezone Europe/London, UTC+1 summer time
        # UTC start_time = 2027-06-06 22:59:59
        # expected start_time_located = 2027-06-06 23:59:59
        # UTC end_time = 2027-06-06 23:00:01
        # expected end_time_located = 2027-06-07 00:00:01
        self.assertFalse(self.shift_utc_plus_1.is_one_day)
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        # UTC end_time = 2027-06-07 00:00:01
+        # expected end_time_located = 2027-06-07 02:00:01
+
+        self.assertTrue(self.shift_utc_plus_2.is_one_day)
```
```suggestion
        # Shift created with timezone Europe/Brussels, UTC+2 summer time
        # UTC start_time = 2027-06-06 22:00:01
        # expected start_time_located = 2027-06-07 00:00:01
        # UTC end_time = 2027-06-07 00:00:01
        # expected end_time_located = 2027-06-07 02:00:01
        self.assertTrue(self.shift_utc_plus_2.is_one_day)
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        self.assertTrue(self.shift_utc_plus_2.is_one_day)
+
+    def test_reduce_max_volunteer_equals_zero(self):
+        """Test that it is possible to reduce max_volunteer_nb to 0"""
```
```suggestion
        """Test that it is not possible to reduce max_volunteer_nb to 0"""
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        under the number of confirmed participation
+        2 participation confirmed, max_volunteer_nb set to 1"""
+
+        self.Participation.create(
```
```suggestion
        """Test that it is not possible to reduce max_volunteer_nb
        under the number of confirmed participation"""
        # Thes is one confirmed participation defined in setUp()
        self.Participation.create(
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        """Test that the remaining slots are correctly computed
+        1 participation confirmed, max_volunteer_nb = 2, remaining_slots = 1"""
+
+        self.assertEqual(self.shift_max_2.remaining_slots, 1)
```
```suggestion
    def test_compute_remaining_slots(self):
        """Test that the remaining slots are correctly computed"""
        # 1 participation confirmed, max_volunteer_nb = 2
        self.assertEqual(self.shift_max_2.remaining_slots, 1)
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        1 participation confirmed, 1 participation canceled,
+        max_volunteer_nb = 2, remaining_slots = 1"""
+
+        self.Participation.create(
```
```suggestion
    def test_compute_remaining_slots_with_canceled_participation(self):
        """Test that the remaining slots are correctly computed
        when there is a canceled participation"""
        # There is one confirmed participation already defined in setUp()
        self.Participation.create(
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift.py:
```diff
@@ -0,0 +1,183 @@
+        1 participation confirmed for the same volunteer,
+        nb_max_volunteer = 2"""
+
+        with self.assertRaises(ValidationError):
```
```suggestion
    def test_volunteer_cannot_confirm_twice_for_same_shift(self):
        """Test that it is not possible to create multiple participation
        for the same volunteer"""
        with self.assertRaises(ValidationError):
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift_participation.py:
```diff
@@ -0,0 +1,30 @@
+from odoo.exceptions import ValidationError
```
```suggestion
# SPDX-FileCopyrightText: 2025 Coop IT Easy SC
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from odoo.exceptions import ValidationError
```

aydrpm replied:
Done

remytms commented on volunteer/tests/test_volunteer_shift_participation.py:
```diff
@@ -0,0 +1,30 @@
+        3 participation confirmed, max_volunteer_nb = 2
+        participation refused"""
+
+        self.Participation.create(
```
```suggestion
    def test_participation_refused_when_shift_full(self):
        """Test that a participation is refused when the number of confirmed
        participation exceeds max_volunteer_nb."""
        # There is a confirmed participation that is already defined in setUp()
        self.Participation.create(
```

aydrpm replied:
Done

remytms commented on volunteer/views/shift_category_view.xml:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/views/shift_kanban_view.xml:
```diff
@@ -0,0 +1,174 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/views/shift_kanban_view.xml:
```diff
@@ -0,0 +1,174 @@
+                                            class="fa fa-arrow-right me-2"
+                                            title="End date"
+                                        />
+                                        <t t-if="record.is_one_day.raw_value == false">
```
```suggestion
                                        <t t-if="!record.is_one_day">
```
Can you try the above suggestion, because comparing a value to `false` is not a good way to do it in general. I think that expression is evaluated in Javascript not in Python.

See also my comment below.

aydrpm replied:
That works, thank you ! 
(Not needed anymore due to the condition swap requested below.)

remytms commented on volunteer/views/shift_kanban_view.xml:
```diff
@@ -0,0 +1,174 @@
+                                                />
+                                            </div>
+                                        </t>
+                                        <t t-else="">
+                                            <span class="o_volunteer_bold">
+                                                <t
+                                                    t-out="luxon.DateTime.fromISO(record.end_time.raw_value).toFormat('HH:mm')"
+                                                />
+                                            </span>
+                                        </t>
```
I think it’s better to put this case first and then the other.

So if is_one_day: simple display
Else: more complex display

It prevent negation in test condition which is easier to read.

aydrpm replied:
Done, thank you.

remytms commented on volunteer/views/shift_kanban_view.xml:
```diff
@@ -0,0 +1,174 @@
+                                        <field name="name" />
+                                    </div>
+
+                                    <t
+                                        t-if="record.tag_ids.raw_value &amp;&amp; record.tag_ids.raw_value.length"
+                                    >
```
```suggestion
                                    <t
                                        t-if="record.tag_ids.raw_value and record.tag_ids.raw_value.length"
                                    >
```
It looks like `if` in kanban views are evaluated in javascript. So normaly using `and` instead of `&&` is totally correct and more readable.

aydrpm replied:
It works, thank you.
It's done.

remytms commented on volunteer/views/shift_kanban_view.xml:
```diff
@@ -0,0 +1,174 @@
+                                                        left
+                                                    </span>
+                                                </div>
+                                            </t>
```
You can dirrectly put the `t-if` condition on the `div` you want to show or hide.

```suggestion
                                            <div
                                                t-if="record.remaining_slots.raw_value &lt;= 0"
                                                class="badge bg-danger o_volunteer_fontsize_11"
                                            >
                                                Full
                                            </div>
                                            <div
                                                t-elif="record.remaining_slots.raw_value == record.max_volunteer_nb.raw_value"
                                                class="badge bg-success o_volunteer_fontsize_11"
                                            >
                                                Empty
                                            </div>
                                            <div
                                                t-elif="record.remaining_slots.raw_value == 1"
                                                class="badge bg-warning"
                                            >
                                                <field
                                                    class="o_volunteer_fontsize_11"
                                                    name="remaining_slots"
                                                />
                                                <span
                                                    class="o_volunteer_fontsize_09"
                                                >
                                                    left
                                                </span>
                                            </div>
                                            <div t-else="" class="badge bg-info">
                                                <field
                                                    class="o_volunteer_fontsize_11"
                                                    name="remaining_slots"
                                                />
                                                <span
                                                    class="o_volunteer_fontsize_09"
                                                >
                                                    left
                                                </span>
                                            </div>
```

aydrpm replied:
Done

remytms commented on volunteer/views/shift_tag_view.xml:
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/views/shift_type_view.xml:
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/views/shift_view.xml:
```diff
@@ -0,0 +1,131 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/views/shift_view.xml:
```diff
@@ -0,0 +1,131 @@
+                <field name="remaining_slots" />
+                <field name="coordinator_id" />
+                <field name="company_id" />
+                <field name="tz" string="Timezone" readonly="1" optional="hide" />
```
Can you move the `tz` field just after the `end_time` field. As it is `hide` by default and related to the date it should be next to the date information.

aydrpm replied:
Done

remytms commented on volunteer/views/volunteer_kanban_view.xml:
```diff
@@ -0,0 +1,55 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/views/volunteer_menu.xml:
```diff
@@ -0,0 +1,78 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms commented on volunteer/views/volunteer_menu.xml:
```diff
@@ -0,0 +1,78 @@
+        id="menu_volunteer_configuration"
+        name="Configuration"
+        parent="menu_volunteer"
+        sequence="40"
```
```suggestion
        sequence="90"
```
You can put a higher sequence for Configuration Menu, as in general we want it last and stay last.

aydrpm replied:
Done

remytms commented on volunteer/views/volunteer_view.xml:
```diff
@@ -0,0 +1,84 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
```
```suggestion
<?xml version="1.0" encoding="utf-8" ?>
<!--
SPDX-FileCopyrightText: 2025 Coop IT Easy SC

SPDX-License-Identifier: AGPL-3.0-or-later
-->
<odoo>
```

aydrpm replied:
Done

remytms reviewed (changes requested):
Great job ! And great first module. :)

General remark, can you specify the name of the arguments when creating fields like this:

```python
foo_id = fields.Many2one(
    comodel_name="res.foo",
    string="Foo",
    ...
)
```
Instead of:
```python
foo_id = fields.Many2one("res.foo", "Foo")
```

See reference [here](https://www.odoo.com/documentation/17.0/developer/reference/backend/orm.html#fields).

I’ve made also specific comment, see other post for details.

remytms commented on volunteer/models/volunteer_shift_participation.py:
```diff
@@ -0,0 +1,69 @@
+
+    # Compute Methods
+
+    @api.depends("registration_state")
+    def _compute_cancellation_date(self):
+        for participation in self:
+            if participation.registration_state == "canceled":
+                participation.cancellation_date = fields.datetime.now()
+            else:
+                participation.cancellation_date = False
```
This cannot be a computed field. You should rather override the `write()` and `create()` method for that. Something like this:

```python
@api.model_create_multi
def create(self, vals_list):
    for vals in vals_list:
        if vals.get("registration_state") == "canceled":
            vals["cancellation_date"] = fields.datetime.now()
    return super().create(vals_list)

def write(self, vals):
    for contact in self:
        old_state = self.registration_state
        new_state = vals.get("registration_state")
        if old_state != "canceled" and new_state == "canceled":
            vals["cancellation_date"] = fields.datetime.now()
    return super().write(new_vals)
```

aydrpm replied:
It's also done, though I'd like to discuss it with you next week if that's okay.

Many thanks for the detailed review and all your suggestions !

aydrpm committed:
Manifest corrections and maintainer addition

aydrpm committed:
Add header Copyright and Licence-Identifier

aydrpm committed:
Fix formatting and docstrings

aydrpm committed:
Add the name of the fields arguments

aydrpm committed:
Minor corrections to shift and menu views

- Moved complex is_one_day condition display to else block (shift
  kanban)
- Replaced "&amp;&amp;" with "and" (shift kanban)
- Placed t-if condition directly on the associated <div> (shift kanban)
- Moved tz field after end_time field (shift tree)
- Increased sequence value for the configuration menu

aydrpm committed:
Replaced search() in _default_stage_id()

Used direct data access instead

aydrpm committed:
Removed unneeded record rules

aydrpm committed:
Handled cancellation_date in write/create

Was previously computed

aydrpm committed:
Remove compute from cancellation_date field

aydrpm committed:
Style: change (4,ref) to (4, ref)

remytms committed:
[CHG] volunteer: version and typo

remytms reviewed (approved):
LGTM. Name of the module may change in next version.

remytms commented:
/ocabot merge nobump
